content/news/2023-10-21-bevy-0.12/index.md

```diff
@@ -19,6 +19,319 @@ Since our last release a few months ago we've added a _ton_ of new features, bug
+    // index into the array and an Option<NonMaxU32> dynamic offset. If storage
+    // buffers are supported, it will be None, else Some with the dynamic
+    // offset that needs to be used when binding the bind group. indices should
+    // be stored somewhere for later lookup, often associated with an Entity.
+    let indices = buffer.push(MyType { x: i as f32 });
+}
+
+// Queue writing the buffer contents to VRAM
+buffer.write_buffer(&render_device, &render_queue);
+
+// The bind group layout entry to use when creating the pipeline
+let binding = 0;
+let visibility = ShaderStages::VERTEX;
+let bind_group_layout_entry = buffer.binding_layout(
+    binding,
+    visibility,
+    &render_device,
+);
+
+// Get the binding resource to make a bind group entry to use when creating the
+// bind group
+let buffer_binding_resource = buffer.binding()?;
+
+// Get the batch size. This will be None if storage buffers are supported, else
+// it is the maximum number of elements that could fit in a batch
+let buffer_batch_size = GpuArrayBuffer::<MyType>::batch_size(&render_device.limits());
+
+// Set a shader def with the buffer batch size
+if let Some(buffer_batch_size) = buffer_batch_size {
+    shader_defs.push(ShaderDefVal::UInt(
+        "BUFFER_BATCH_SIZE".into(),
+        buffer_batch_size,
+    ));
+}
+```
+
+```rust
+#import bevy_render::instance_index get_instance_index
+
+struct MyType {
+    x: f32,
+}
+
+// Declare the buffer binding
+#ifdef BUFFER_BATCH_SIZE
+@group(2) @binding(0) var<uniform> data: array<MyType, #{BUFFER_BATCH_SIZE}u>;
+#else
+@group(2) @binding(0) var<storage> data: array<MyType>;
+#endif
+
+// Access an instance
+let my_type = data[get_instance_index(in.instance_index)];
+```
+
+### 2D / 3D Mesh Entities using GpuArrayBuffer
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @robtfm, @Elabajaba</div>
+
+The 2D and 3D mesh entity rendering was migrated to use `GpuArrayBuffer` for the mesh uniform data.
+
+Just avoiding the rebinding of the mesh uniform data buffer gives about a 6% increase in frame rates.
+
+### Improved bevymark Example
+
+<div class="release-feature-authors">authors: Rob Swain (@superdump), @IceSentry</div>
+
+The bevymark example needed to be improved to enable benchmarking the batching / instanced draw changes. Modes were added to:
+
+* draw 2D quad meshes instead of sprites: `--mode mesh2d`
+* vary the per-instance color data instead of only varying the colour per wave of birds: `--vary-per-instance`
+* generate a number of material / sprite textures and randomly choose from them either per wave or per instance depending on the vary per instance setting: `--material-texture-count 10`
+* spawn the birds in random z order (new default), or in draw order: `--ordered-z`
+
+This allows benchmarking of different situations for batching / instancing in the next section.
+
+### Automatic Batching/Instancing of Draw Commands
```

Comment:
I think we should move this section higher up. I know it builds on other things, but it feels like we're burying the lead here. There are ways to reference an abstraction used prior to defining it.